File: 0746-min-cost-climbing-stairs/0746-min-cost-climbing-stairs.java

class Solution {
    int rec(int i, int[] cost, int[] memo) {
        if (i <= 1) return 0;
        if (memo[i] != -1) return memo[i];

        int oneStep = rec(i - 1, cost, memo) + cost[i - 1];
        int twoStep = rec(i - 2, cost, memo) + cost[i - 2];

        return memo[i] = Math.min(oneStep, twoStep);
    }

    public int minCostClimbingStairs(int[] cost) {
        int n = cost.length;
        int[] memo = new int[n + 1];
        Arrays.fill(memo, -1);
        memo[0] = memo[1] = 0;
        return rec(n, cost, memo);
    }
}

File: 0746-min-cost-climbing-stairs/0746-min-cost-climbing-stairs.py
class Solution:
    def minCostClimbingStairs(self, cost: List[int]) -> int:
        dp=[-1]*len(cost)
        ans1=self.solve(cost,0,dp)
        ans2=self.solve(cost,1,dp)
        return min(ans1,ans2)

    def solve(self,cost,i,dp):
        if i >= len(cost):
            return 0
        if dp[i]!=-1:
            return dp[i]
        
        dp[i]=min(cost[i]+self.solve(cost,i+1,dp),cost[i]+self.solve(cost,i+2,dp))
        return dp[i]